docs/DesignDocs/Protected_Code_Loader_Proposal.md

```diff
@@ -0,0 +1,161 @@
+enclave image with the compilation option "-Wl, whole-archive"
+### Encrypt the enclave image
+An encryption tool is provided as the pre-step to enclave signing process:
+
+```bash
+oeencrypt -i my_enclave -o my_enclave.enc -k keyfile
+```
+
+## Enclave Load-time:
+On Enclave Loading, if the image is encrypted, it need to be decrypted 
+inside enclave loader, this step is done by OE SDK runtime and invisible 
+to user. 
+
+# Specification
+
+## Protected Code Loader Software Work Flow in Open Enclave
+### Build time encryption 
+    A seperate tool "oeencrypt" is provided at enclave build time to encrypt 
+	the enclave image right before enclave signing process.  
+### Load time decryption
+    On enclave initliazation operation(EINIT operation), right before relocation, 
+    the encrypted enclave image need to be decrypted and than perform the relocation
+    operation.
+
+## Encryption Algorithm in Protected Code Loader
+Using openssl AES-256-GCM as the encryption/decryption algorithm.
+
+## ELF sections Left plaintext
+### PCL table entry 
+A PCL table entry (a section called ".pcltbl" in ELF file)is built into the Enclave 
+image by linking to PCL lib, i.e. liboepcl.a, this part contains enclave decryption
+info on enclave loading and must remain plaintext.
+
+Definition of PCL entry(section ".pcltbl"):
+typedef struct pcl_table_t_
+{
+	/* Current state of PCL: initailized to PCL_PLAIN */
+    pcl_status_e pcl_state;                   
+    uint32_t     reserved1[3];                /* Must be 0 */
+	// GUID must match GUID in Sealed blob
+    uint8_t      pcl_guid[SGX_PCL_GUID_SIZE]; 
+    size_t       sealed_blob_size;            
+    uint32_t     reserved2[2];                /* Must be 0 */
+	/* For security, sealed blob is copied into enclave */
+    uint8_t      sealed_blob[PCL_SEALED_BLOB_SIZE];
+	/* SHA256 digest of decryption key */
+    uint8_t      decryption_key_hash[SGX_SHA256_HASH_SIZE];
+	/* Number of RVAs */
+    uint32_t     num_rvas;                    
+    uint32_t     reserved3[3];                /* Must be 0 */
+	/* Array of rva_size_tag_iv_t */
+    rva_size_tag_iv_t rvas_sizes_tags_ivs[PCL_MAX_NUM_ENCRYPTED_SECTIONS]; 
+}pcl_table_t;
+
+### Sections left plaintext
+1. ELF header  - binary header
+2. Sections table
+3. Segments table
+4. Sections' names string table pointed by e_shstrndx(e.g. .shstrtab)
+5. .oeinfo section holds enclave's metadata(properties)
+6. .bss and .tbss
+7. sections required to construct dyn_info (.dynamic)
+8. sections holds the content provided by entries with index DT_SYMTAB, DT_STRTAB and DT_REL in
+   dyn_info (e.g. .dynsym, .dynstr, .rela.dyn)
+9. sections containing PCL code and data:
+   a. section ".pcltbl"  // Designated section for PCL table
+   b. .nipx, .nipd, .niprod, .nipd_rel, .nipd_rel_ro_local
+10. sections for debugging
+   .comment, .debug_abbrev, .debug_aranges, .debug_info, .debug_line, .debug_lc, .debug_ranges,
+   .debug_str
+
+## Elf sections to be encrypted
```

Comment:
For consistency, capitalize the first character of words that are not articles/prepositions in title/sub-title.

```suggestion
## Elf Sections to Be Encrypted
```